### PR TITLE
MediaPlayerPrivateWebM caused the seeked event to be fired before the seek has completed.

### PR DESCRIPTION
**Pull request by jyavenard**
#### 925970b7979d9517eebb9014b76ef399c9572f4a
<pre>
MediaPlayerPrivateWebM caused the seeked event to be fired before the seek has completed.
<a href="https://bugs.webkit.org/show_bug.cgi?id=267326">https://bugs.webkit.org/show_bug.cgi?id=267326</a>
<a href="https://rdar.apple.com/120776794">rdar://120776794</a>

Reviewed by Youenn Fablet.

MediaPlayerPrivateWebM only implemented the most basic of seeking operation.
We copy the logic from the MediaPlayerPrivateMediaSourceAVFObjC into the WebM
player.

Now the seeked event will only be fired once the seek operation has completed
and once the frame at the seek operation has completed.
The WebM player would also always start playback after a seek regardless
of the element playing prior the operation, only resume playback if we
were seeking before.

Following a call to flush() WebCoreDecompressionSession could continue
delivering in-flight decompressed frame. We modify it so that we can guarantee
that no frames enqueued before the call to flush() can get delivered.
This is done by using NativePromise to track the actual decoding and storing
the flushId value at the start of the decoding, and to stop delivering the
frame if that flushId has changed.

Add tests.
On macOS &lt; 14 and iOS &lt; 17; the required `isReadyForDisplay` API to accurately
determine that a frame got painted doesn&apos;t exist.

* LayoutTests/media/content/test-vp8.webm: Added.
* LayoutTests/media/content/test-vp8.webm.png: Added.
* LayoutTests/media/media-vp8-webm-expected.html: Added.
* LayoutTests/media/media-vp8-webm.html: Added.
* LayoutTests/platform/mac-wk1/TestExpectations:
* Source/WebCore/platform/graphics/cocoa/MediaPlayerPrivateWebM.h:
* Source/WebCore/platform/graphics/cocoa/MediaPlayerPrivateWebM.mm:
(WebCore::MediaPlayerPrivateWebM::MediaPlayerPrivateWebM):
(WebCore::MediaPlayerPrivateWebM::~MediaPlayerPrivateWebM):
(WebCore::MediaPlayerPrivateWebM::play):
(WebCore::MediaPlayerPrivateWebM::pause):
(WebCore::MediaPlayerPrivateWebM::seekToTarget):
(WebCore::MediaPlayerPrivateWebM::seekInternal):
(WebCore::MediaPlayerPrivateWebM::maybeCompleteSeek):
(WebCore::MediaPlayerPrivateWebM::seeking const):
(WebCore::MediaPlayerPrivateWebM::clampTimeToLastSeekTime const):
(WebCore::MediaPlayerPrivateWebM::shouldBePlaying const):
(WebCore::MediaPlayerPrivateWebM::setRateDouble):
(WebCore::MediaPlayerPrivateWebM::setHasAvailableVideoFrame):
(WebCore::MediaPlayerPrivateWebM::enqueueSample):
(WebCore::MediaPlayerPrivateWebM::destroyDecompressionSession):
(WebCore::MediaPlayerPrivateWebM::layerDidReceiveError):
(WebCore::MediaPlayerPrivateWebM::layerReadyForDisplayChanged):
* Source/WebCore/platform/graphics/cocoa/WebCoreDecompressionSession.h: Remove unused m_consumerQueue related code
* Source/WebCore/platform/graphics/cocoa/WebCoreDecompressionSession.mm:
(WebCore::WebCoreDecompressionSession::enqueueSample):
(WebCore::WebCoreDecompressionSession::enqueueCompressedSample):
(WebCore::WebCoreDecompressionSession::maybeDecodeNextSample):
(WebCore::WebCoreDecompressionSession::decodeSample):
(WebCore::WebCoreDecompressionSession::handleDecompressionOutput):
(WebCore::WebCoreDecompressionSession::isReadyForMoreMediaData const):
(WebCore::WebCoreDecompressionSession::flush):
(WebCore::WebCoreDecompressionSession::finishCurrentDecodingAndReportError): Deleted.
(WebCore::WebCoreDecompressionSession::finishCurrentDecodingAndDecodeNextSample): Deleted.

Canonical link: <a href="https://commits.webkit.org/273109@main">https://commits.webkit.org/273109@main</a>
</pre>

https://github.com/WebKit/WebKit/commit/51ff829a73540de2f465679d37bc2954f57f7638

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/34290 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/13098 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/36282 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/36971 "Built successfully") | [  ~~🛠 wincairo~~](https://ews-build.webkit.org/#/builders/32/builds/31049 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/35372 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/15481 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/51/builds/10220 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/34/builds/30064 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/34802 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/11090 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/30593 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/9678 "Passed tests") | 
| | [  ~~🧪 ios-wk2-wpt~~](https://ews-build.webkit.org/#/builders/42/builds/9779 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/30639 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/38259 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/31153 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/30960 "Passed tests") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/1/builds/35854 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| | [  ~~🛠 tv~~](https://ews-build.webkit.org/#/builders/44/builds/9904 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/50/builds/7772 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/33779 "Passed tests") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/11699 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/30263 "Passed tests") | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/7888 "Built successfully and passed tests") | [  ~~🛠 watch~~](https://ews-build.webkit.org/#/builders/43/builds/10461 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/10735 "Built successfully") | | |

### CODE REVIEW

File: Source/WebCore/platform/graphics/cocoa/MediaPlayerPrivateWebM.h [hunk 1]
```diff
@@ -328,6 +326,27 @@ class MediaPlayerPrivateWebM
     bool m_errored { false };
     bool m_processingInitializationSegment { false };
     Ref<WebAVSampleBufferListener> m_listener;
+
+    // Seek logic support
+    void seekToTarget(const SeekTarget&) final;
```
Comment (youennf) [inline Source/WebCore/platform/graphics/cocoa/MediaPlayerPrivateWebM.h:331]: I would put these methods above to not split the members in two sections.

File: Source/WebCore/platform/graphics/cocoa/MediaPlayerPrivateWebM.mm [hunk 2]
```diff
@@ -89,9 +89,32 @@ @interface AVSampleBufferDisplayLayer (Staging_100128644)
...
     m_parser->setLogger(m_logger, m_logIdentifier);
+
+    // addPeriodicTimeObserverForInterval: throws an exception if you pass a non-numeric CMTime, so just use
+    // an arbitrarily large time value of once an hour:
+    __block WeakPtr weakThis { *this };
+    m_timeJumpedObserver = [m_synchronizer addPeriodicTimeObserverForInterval:PAL::toCMTime(MediaTime::createWithDouble(3600)) queue:dispatch_get_main_queue() usingBlock:^(CMTime time) {
+#if LOG_DISABLED
+        UNUSED_PARAM(time);
+#endif
+        // FIXME: Remove the below once <rdar://problem/15798050> is fixed.
```
Comment (youennf) [inline Source/WebCore/platform/graphics/cocoa/MediaPlayerPrivateWebM.mm]: It seems fixed in 2014, I am unclear what is needed here.

File: Source/WebCore/platform/graphics/cocoa/MediaPlayerPrivateWebM.mm [hunk 3]
```diff
@@ -89,9 +89,32 @@ @interface AVSampleBufferDisplayLayer (Staging_100128644)
     , m_logIdentifier(player->mediaPlayerLogIdentifier())
     , m_videoLayerManager(makeUnique<VideoLayerManagerObjC>(m_logger, m_logIdentifier))
     , m_listener(WebAVSampleBufferListener::create(*this))
+    , m_seekTimer(*this, &MediaPlayerPrivateWebM::seekInternal)
 {
     ALWAYS_LOG(LOGIDENTIFIER);
     m_parser->setLogger(m_logger, m_logIdentifier);
+
+    // addPeriodicTimeObserverForInterval: throws an exception if you pass a non-numeric CMTime, so just use
+    // an arbitrarily large time value of once an hour:
```
Comment (youennf) [inline Source/WebCore/platform/graphics/cocoa/MediaPlayerPrivateWebM.mm:98]: Why is an hour large enough?
Does it mean we might have an issue every one hour or so?
I would tend to put something bigger then, maybe 24 hours.

File: Source/WebCore/platform/graphics/cocoa/MediaPlayerPrivateWebM.mm [hunk 4]
```diff
@@ -262,41 +292,113 @@ static bool isCopyDisplayedPixelBufferAvailable()
 
 MediaTime MediaPlayerPrivateWebM::currentMediaTime() const
 {
-    MediaTime synchronizerTime = PAL::toMediaTime(PAL::CMTimebaseGetTime([m_synchronizer timebase]));
+    MediaTime synchronizerTime = clampTimeToLastSeekTime(PAL::toMediaTime(PAL::CMTimebaseGetTime([m_synchronizer timebase])));
     if (synchronizerTime < MediaTime::zeroTime())
         return MediaTime::zeroTime();
-    if (synchronizerTime > durationMediaTime())
-        return durationMediaTime();
```
Comment (youennf) [inline Source/WebCore/platform/graphics/cocoa/MediaPlayerPrivateWebM.mm:269]: Why is it no longer needed?

File: Source/WebCore/platform/graphics/cocoa/MediaPlayerPrivateWebM.mm [hunk 5]
```diff
@@ -262,41 +292,113 @@ static bool isCopyDisplayedPixelBufferAvailable()
...
+    if (synchronizerTime < m_lastSeekTime)
```
Comment (youennf) [inline Source/WebCore/platform/graphics/cocoa/MediaPlayerPrivateWebM.mm]: Isn't it redundant with what clampTimeToLastSeekTime does?

File: Source/WebCore/platform/graphics/cocoa/MediaPlayerPrivateWebM.h [hunk 6]
```diff
@@ -328,6 +326,27 @@ class MediaPlayerPrivateWebM
...
+    Timer m_seekTimer;
+    MediaTime m_lastSeekTime;
+    std::optional<SeekTarget> m_pendingSeek;
+    enum SeekState {
+        Seeking,
+        WaitingForAvailableFame,
+        SeekCompleted,
+    };
+    SeekState m_seekState { SeekCompleted };
+    bool m_synchronizerSeeking { false };
```
Comment (youennf) [inline Source/WebCore/platform/graphics/cocoa/MediaPlayerPrivateWebM.h]:
```suggestion
    bool m_isSynchronizerSeeking { false };
```

File: Source/WebCore/platform/graphics/cocoa/MediaPlayerPrivateWebM.h [hunk 7]
```diff
@@ -328,6 +326,27 @@ class MediaPlayerPrivateWebM
...
+
+    // Seek logic support
+    void seekToTarget(const SeekTarget&) final;
+    bool seeking() const final;
+    void seekInternal();
+    void maybeCompleteSeek();
+    MediaTime clampTimeToLastSeekTime(const MediaTime&) const;
+    bool shouldBePlaying() const;
+
+    bool m_playing { false };
```
Comment (youennf) [inline Source/WebCore/platform/graphics/cocoa/MediaPlayerPrivateWebM.h]:
```suggestion
    bool m_isPlaying { false };
```

File: Source/WebCore/platform/graphics/cocoa/MediaPlayerPrivateWebM.mm [hunk 8]
```diff
@@ -262,41 +292,113 @@ static bool isCopyDisplayedPixelBufferAvailable()
...
     if (auto player = m_player.get()) {
-        player->seeked(target.time);
+        player->seeked(m_lastSeekTime);
         player->timeChanged();
     }
 }
 
+bool MediaPlayerPrivateWebM::seeking() const
+{
+    return m_pendingSeek || m_synchronizerSeeking || m_seekState != SeekCompleted;
```
Comment (youennf) [inline Source/WebCore/platform/graphics/cocoa/MediaPlayerPrivateWebM.mm]: It is not totally clear why we need m_synchronizerSeeking given we have m_seekState.
Can we remove m_synchronizerSeeking?

File: Source/WebCore/platform/graphics/cocoa/WebCoreDecompressionSession.mm [hunk 9]
```diff
@@ -287,103 +266,156 @@
...
+            if (*result) {
+                if (!m_deliverDecodedFrames) {
+                    m_enqueingQueue->dispatch([protectedThis = Ref { *this }, imageSampleBuffer = WTFMove(*result), flushId] {
+                        if (flushId == protectedThis->m_flushId)
+                            protectedThis->enqueueDecodedSample(imageSampleBuffer.get());
+                    });
+                } else {
+                    RunLoop::main().dispatch([protectedThis = Ref { *this }, this, imageSampleBuffer = WTFMove(*result), flushId]() mutable {
+                        assertIsMainThread();
+                        if (flushId == m_flushId && m_newDecodedFrameCallback) {
```
Comment (youennf) [inline Source/WebCore/platform/graphics/cocoa/WebCoreDecompressionSession.mm:298]: Why can m_flushId be checked from either main thread or m_enqueingQueue?

File: Source/WebCore/platform/graphics/cocoa/MediaPlayerPrivateWebM.h (hunk 1 above)
Comment (jyavenard) [reply Source/WebCore/platform/graphics/cocoa/MediaPlayerPrivateWebM.h:331]: I like having the seek related in their own section better. There's two many virtual methods, it becomes hard to follow

File: Source/WebCore/platform/graphics/cocoa/MediaPlayerPrivateWebM.mm (hunk 2 above)
Comment (jyavenard) [reply Source/WebCore/platform/graphics/cocoa/MediaPlayerPrivateWebM.mm]: This change replicate the well tested MediaPlayerPrivateMediaSourceAVFObjC 

Idea is to replicate the code that is known to work.

I don't take the assumption that what was currently there was correct, because it wasn't.

File: Source/WebCore/platform/graphics/cocoa/MediaPlayerPrivateWebM.mm (hunk 3 above)
Comment (jyavenard) [reply Source/WebCore/platform/graphics/cocoa/MediaPlayerPrivateWebM.mm:98]: As above, this is replicating the code from `MediaPlayerPrivateMediaSourceAVFObjC`

if this is to be removed, I will do so as a follow-up and remove it wherever it's left

File: Source/WebCore/platform/graphics/cocoa/MediaPlayerPrivateWebM.mm (hunk 4 above)
Comment (jyavenard) [reply Source/WebCore/platform/graphics/cocoa/MediaPlayerPrivateWebM.mm:269]: It was never needed. durationMediaTime is the value of max(duration, currentTime) as set by the observer.

It is a copy of MediaPlayerPrivateMediaSourceAVFObjC code

File: Source/WebCore/platform/graphics/cocoa/MediaPlayerPrivateWebM.mm (hunk 5 above)
Comment (jyavenard) [reply Source/WebCore/platform/graphics/cocoa/MediaPlayerPrivateWebM.mm]: Probably, but again, this is copying the logic from MediaPlayerPrivateMediaSourceAVFObjC and I don't want to modify MediaPlayerPrivateMediaSourceAVFObjC at this late stage.

File: Source/WebCore/platform/graphics/cocoa/MediaPlayerPrivateWebM.mm (hunk 8 above)
Comment (jyavenard) [reply Source/WebCore/platform/graphics/cocoa/MediaPlayerPrivateWebM.mm]: It would mean to redo/review all the existing logic , I don't want to do it at this stage.

File: Source/WebCore/platform/graphics/cocoa/WebCoreDecompressionSession.mm (hunk 9 above)
Comment (jyavenard) [reply Source/WebCore/platform/graphics/cocoa/WebCoreDecompressionSession.mm:298]: m_flushId is an atomic
it all depends on where the end result occurs.
with decodedFrameCallback ; it's called on the main thread, so this is where flushId is checked against.

For the enqueuing of frames, this occurs on the enqueuing thread, and so this is where flushed is checked against.

File: Source/WebCore/platform/graphics/cocoa/WebCoreDecompressionSession.mm [hunk 10]
```diff
@@ -287,103 +266,156 @@
...
+            });
+        } else {
+            if (*result) {
+                if (!m_deliverDecodedFrames) {
+                    m_enqueingQueue->dispatch([protectedThis = Ref { *this }, imageSampleBuffer = WTFMove(*result), flushId] {
+                        if (flushId == protectedThis->m_flushId)
+                            protectedThis->enqueueDecodedSample(imageSampleBuffer.get());
+                    });
+                } else {
+                    RunLoop::main().dispatch([protectedThis = Ref { *this }, this, imageSampleBuffer = WTFMove(*result), flushId]() mutable {
```
Comment (youennf) [inline Source/WebCore/platform/graphics/cocoa/WebCoreDecompressionSession.mm]: ensureOnMainThread, given we are using ensureOnMainThread just a few lines above.

**Review by youennf — APPROVED** [review]
r=me.
Given this code is copy and paste from the MSE player, it begs the question of sharing some code between the two, otherwise we might fix one code path and not the other.
This can be done as a follow-up though, can you file a bug to keep track of this idea?

I would fix the few nits below, it does not seem good to keep code that we think is not useful.

File: Source/WebCore/platform/graphics/cocoa/MediaPlayerPrivateWebM.mm (hunk 8 above)
Comment (jyavenard) [reply Source/WebCore/platform/graphics/cocoa/MediaPlayerPrivateWebM.mm]: ok, I misread your comment, I thought you were referring to the entire code, not just this method. yes will do
